### PR TITLE
Added Vim segment plugin for the Asynchronous Linter Engine (ALE)

### PR DESCRIPTION
**Pull request by pmacosta**
This segment has the same functionality and interface as the Syntastic
Vim segment plugin, and the code is largely adapted from it

### CODE REVIEW

File: powerline/segments/vim/plugin/ale.py [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+	:param str err_format:
+		Format string for errors.
+
+	:param str warn_format:
+		Format string for warnings.
+
+	Highlight groups used: ``ale:warning`` or
+	``warning``, ``ale:error`` or ``error``.
+	"""
+	# pylint: disable=C0103,W0613
```
Comment (ZyX-I) [inline powerline/segments/vim/plugin/ale.py]: Do not add `pylint` comments, powerline is not using it.

File: powerline/segments/vim/plugin/ale.py [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+	debug = False
+	bufnr = str(segment_info['bufnr'])
+	ale_enabled_exists = vim_global_exists('ale_enabled')
+	ale_enabled = ale_enabled_exists and int(vim.eval('g:ale_enabled'))
+	if debug:
```
Comment (ZyX-I) [inline powerline/segments/vim/plugin/ale.py]: This should not be hidden under `if debug`, you should use `pl.debug` instead.

File: powerline/segments/vim/plugin/ale.py [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+		pl.info('ale_enabled_exists: {0}'.format(ale_enabled_exists))
```
Comment (ZyX-I) [inline powerline/segments/vim/plugin/ale.py]: Do not use `.format`, `PowerlineLogger` does already use it internally.

File: powerline/segments/vim/plugin/ale.py [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+#	  "log_level": "INFO"
+# },
+#
+# and then turn on debug internal variable to True
+@requires_segment_info
+def ale(
+	segment_info,
+	pl,
+	err_format='ERR:  {first_line} ({num}) ',
+	warn_format='WARN:  {first_line} ({num}) '
```
Comment (ZyX-I) [inline powerline/segments/vim/plugin/ale.py]: Should not use non-ASCII here any longer, actually I need to remove that from syntastic as well. All non-ASCII should go to relevant top-level theme files like `/powerline/config_files/themes/powerline.json`.

File: powerline/segments/vim/plugin/ale.py [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+	has_errors = int(vim.eval(cmd))
+	if debug:
+		pl.info('has_errors: {0}'.format(has_errors))
+	if not has_errors:
+		return
+	cmd = 'ale#engine#GetLoclist({0})'.format(bufnr)
+	if debug:
+		pl.info('cmd: {0}'.format(cmd))
+	issues = vim.eval(cmd)
+	errors, warnings = [], []
```
Comment (ZyX-I) [inline powerline/segments/vim/plugin/ale.py]: Do not have multiple definitions on one line unless you have to (i.e. when unpacked tuple returned by the function). Also do not use lists, you need *one* entry of the whole list and list length after all, you don’t need the entire list.

File: powerline/segments/vim/plugin/ale.py [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
+	or warnings
+
+	:param str err_format:
+		Format string for errors.
+
+	:param str warn_format:
+		Format string for warnings.
+
+	Highlight groups used: ``ale:warning`` or
+	``warning``, ``ale:error`` or ``error``.
```
Comment (ZyX-I) [inline powerline/segments/vim/plugin/ale.py]: Must not wrap, this comment is here for the *linter* so the format needs to be *exact*, like described in develop/segments.rst.

File: powerline/segments/vim/plugin/ale.py [hunk 7]
```diff
@@ -0,0 +1,97 @@
...
+# and then turn on debug internal variable to True
+@requires_segment_info
+def ale(
+	segment_info,
+	pl,
+	err_format='ERR:  {first_line} ({num}) ',
+	warn_format='WARN:  {first_line} ({num}) '
+):
+	"""
+	Show whether Asynchronous Lint Engine (ALE) has found any errors
```
Comment (ZyX-I) [inline powerline/segments/vim/plugin/ale.py]: Too long summary, summary not on the same line with quotes, using double quotes in place of single.

File: powerline/segments/vim/plugin/ale.py [hunk 8]
```diff
@@ -0,0 +1,97 @@
+# vim:fileencoding=utf-8:noet
+"""
+Asynchronous Lint Engine (ALE) plugin
+
+Largely adapted from Sytastic plugin
+"""
+
+from __future__ import (
+	unicode_literals, division, absolute_import, print_function
+)
```
Comment (ZyX-I) [inline powerline/segments/vim/plugin/ale.py]: Just copy the header from syntastic: no wrapping, more empty lines and no docstring.

Comment (pmacosta) [reply powerline/segments/vim/plugin/ale.py]: Fixed in new commit

File: powerline/segments/vim/plugin/ale.py (hunk 4 above)
Comment (pmacosta) [reply powerline/segments/vim/plugin/ale.py]: Fixed in new commit, file is ASCII-encoded now

File: powerline/segments/vim/plugin/ale.py (hunk 7 above)
Comment (pmacosta) [reply powerline/segments/vim/plugin/ale.py]: Fixed in new commit

File: powerline/segments/vim/plugin/ale.py (hunk 6 above)
Comment (pmacosta) [reply powerline/segments/vim/plugin/ale.py]: Fixed in new commit

File: powerline/segments/vim/plugin/ale.py (hunk 1 above)
Comment (pmacosta) [reply powerline/segments/vim/plugin/ale.py]: Fixed in new commit

File: powerline/segments/vim/plugin/ale.py (hunk 3 above)
Comment (pmacosta) [reply powerline/segments/vim/plugin/ale.py]: Actually removed all debug logging, in hindsight this was only needed to get the ALE interfacing correct, and it only clutters code now. Removal also makes Syntastic and ALE plugin files more similar.

File: powerline/segments/vim/plugin/ale.py (hunk 5 above)
Comment (pmacosta) [reply powerline/segments/vim/plugin/ale.py]: Location list is traversed once, first error and warning are captured and from then on only error and warning count is kept

File: powerline/segments/vim/plugin/ale.py (hunk 2 above)
Comment (pmacosta) [reply powerline/segments/vim/plugin/ale.py]: Actually removed all debug logging, in hindsight this was only needed to get the ALE interfacing correct, and it only clutters code now. Removal also makes Syntastic and ALE plugin files more similar.

File: powerline/segments/vim/plugin/ale.py (hunk 3 above)
Comment (pmacosta) [reply powerline/segments/vim/plugin/ale.py]: Removed all logging code

File: powerline/segments/vim/plugin/ale.py [hunk 9]
```diff
@@ -0,0 +1,51 @@
+from __future__ import (unicode_literals, division, absolute_import, print_function)
```
Comment (ZyX-I) [inline powerline/segments/vim/plugin/ale.py]: Now it is missing modeline. LGTM, except for this and lack of spaces around some operators (`+`) issues.

Comment (pmacosta) [reply powerline/segments/vim/plugin/ale.py]: Sorry, brought back modeline and added spaces around plus signs

Comment (pmacosta) [reply powerline/segments/vim/plugin/ale.py]: Happy to squash the commits if that is your preference

**Pull request merged**